stdlib/public/core/Stride.swift.gyb

```diff
@@ -487,93 +442,41 @@ public struct StrideThrough<
   }
 }
 
-// FIXME(conditional-conformances): these extra types can easily be turned into
-// conditional extensions to StrideThrough type
-% for Self, ElementConstraint, Where in [
-%   ('IntegerStrideThroughCollection', 'BinaryInteger', 'Element.Stride : BinaryInteger'),
-%   ('FloatingPointStrideThroughCollection', 'BinaryFloatingPoint', 'Element.Stride == Element'),
-% ]:
-%   ElementIsInteger = ElementConstraint == 'BinaryInteger'
-
-internal struct ${Self}<
-  Element : ${ElementConstraint}
-> : RandomAccessCollection, CustomReflectable
-where ${Where} {
-
-//===----------------------------------------------------------------------===//
-// This block is copied from StrideThrough struct definition                  //
-//===----------------------------------------------------------------------===//
-  /// Returns an iterator over the elements of this sequence.
-  ///
-  /// - Complexity: O(1).
-  @_inlineable
-  public func makeIterator() -> StrideThroughIterator<Element> {
-    return StrideThroughIterator(_start: _start, end: _end, stride: _stride)
-  }
-
-  @_inlineable
-  public func _customContainsEquatableElement(
-    _ element: Element
-  ) -> Bool? {
-    if element < _start || _end < element {
-      return false
-    }
-    return nil
-  }
-
-  @_inlineable
-  @_versioned
-  internal init(_start: Element, end: Element, stride: Element.Stride) {
-    _precondition(stride != 0, "Stride size must not be zero")
-    self._start = _start
-    self._end = end
-    self._stride = stride
-  }
-
-  @_versioned
-  internal let _start: Element
-  @_versioned
-  internal let _end: Element
-  @_versioned
-  internal let _stride: Element.Stride
-
-  @_inlineable // FIXME(sil-serialize-all)
-  public var customMirror: Mirror {
-    return Mirror(self,
-      children: ["from": _start, "through": _end, "by": _stride])
-  }
-//===----------------------------------------------------------------------===//
-// The end of the copied block
-//===----------------------------------------------------------------------===//
-
-  // RandomAccessCollection conformance
+// FIXME(conditional-conformances): This does not yet compile (SR-6474).
+#if false
+extension StrideThrough : RandomAccessCollection
+where Element.Stride : BinaryInteger {
   public typealias Index = ClosedRangeIndex<Int>
   public typealias IndexDistance = Int
-  public typealias SubSequence = RandomAccessSlice<${Self}>
+  public typealias SubSequence = RandomAccessSlice<StrideThrough<Element>>
 
   @_inlineable
-  public var startIndex: Index { return ClosedRangeIndex(0) }
+  public var startIndex: Index {
+    let distance = _start.distance(to: _end)
+    return distance == 0 || (distance < 0) == (_stride < 0)
+      ? ClosedRangeIndex(0)
+      : ClosedRangeIndex()
```

Comment:
I think we should just disallow creating strides that don't make sense. `StrideThrough` cannot be empty and that should be enforced in the initializer. Also the `(distance < 0) == (_stride < 0)` check, I think, should be performed in the initializer. These ill-formed sequences/collections don't seem useful to me.

Reply:
If I had to design it from scratch, I absolutely agree. But it’d be source-breaking and there’s nothing that’s really broken by continuing to support it.

Reply:
Source breaking? Nah... it's just a runtime breakage.
But seriously, if anyone is using the empty stride-through or a mis-directed one, their code is already broken.
`stride(from:1, through:0, by:1)` should be similar to `1...0` and it's not.
Let me try to introduce the checks and run the compat test suite.

Reply:
I would be wary of adding a new precondition for the `stride` functions, since we've never had `start <= end` documented as a requirement. I wouldn't at all be surprised if people have switched from something like `n...count` to `stride(from: n, through: count, by: 1)` specifically to avoid having to check for `n > count`.

Reply:
Precisely.

Reply:
I'm not saying we should just do it. It should go through evolution and such. It feels wrong to me allowing incorrect ranges in one place but not the other. As for the check, it will happen either way, if not explicitly in the user code, then in the iterator, that will just return `nil`s all the  time.